File: app/src/release/java/org/tasks/BuildSetup.java
package org.tasks;

import android.util.Log;
import javax.inject.Inject;
import timber.log.Timber;

public class BuildSetup {

  @Inject
  public BuildSetup() {}

  public boolean setup() {
    Timber.plant(new ErrorReportingTree());
    return true;
  }

  private static class ErrorReportingTree extends Timber.Tree {

    @Override
    protected void log(int priority, String tag, String message, Throwable t) {
      if (priority < Log.WARN) {
        return;
      }
      if (priority == Log.ERROR) {
        if (t == null) {
          Log.e(tag, message);
        } else {
          Log.e(tag, message, t);
        }
      } else if (priority == Log.WARN) {
        if (t == null) {
          Log.w(tag, message);
        } else {
          Log.w(tag, message, t);
        }
      }
    }
  }
}

File: app/src/release/java/org/tasks/BuildSetup.kt
package org.tasks

import android.annotation.SuppressLint
import android.app.Application
import android.util.Log
import org.tasks.logging.FileLogger
import timber.log.Timber
import javax.inject.Inject

class BuildSetup @Inject constructor(
    private val context: Application,
) {
    fun setup() {
        Timber.plant(ErrorReportingTree())
        Timber.plant(FileLogger(context))
    }

    private class ErrorReportingTree : Timber.Tree() {
        @SuppressLint("LogNotTimber")
        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
            if (priority < Log.WARN) {
                return
            }
            if (priority == Log.ERROR) {
                if (t == null) {
                    Log.e(tag, message)
                } else {
                    Log.e(tag, message, t)
                }
            } else if (priority == Log.WARN) {
                if (t == null) {
                    Log.w(tag, message)
                } else {
                    Log.w(tag, message, t)
                }
            }
        }
    }
}